Sources/NIOCore/IPProtocol.swift

```diff
@@ -0,0 +1,338 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the SwiftNIO open source project
+//
+// Copyright (c) 2022 Apple Inc. and the SwiftNIO project authors
+// Licensed under Apache License v2.0
+//
+// See LICENSE.txt for license information
+// See CONTRIBUTORS.txt for the list of SwiftNIO project authors
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+//===----------------------------------------------------------------------===//
+
+/// In the Internet Protocol version 4 (IPv4) [RFC791] there is a field
+/// called "Protocol" to identify the next level protocol.  This is an 8
+/// bit field.  In Internet Protocol version 6 (IPv6) [RFC8200], this field
+/// is called the "Next Header" field.
+public struct NIOIPProtocol: RawRepresentable, Hashable {
```

Comment:
For the sake of convenience it might be nice to add a `CustomStringConvertible` implementation that switches over the named `static let`s and prints those names as well as the raw number.